error: ‘Empty’ was not declared in this scope

In file included from /home/simple/code/saas/ae-arbiter/executables/yakoo-server.cc:7:
/home/simple/code/saas/ae-arbiter/src/grpcxx/predictor.grpcxx.pb.h:17:43: error: ‘Empty’ was not declared in this scope
   17 | using rpcversion = grpcxx::rpc<"version", Empty, VersionResponse>;
      |                                           ^~~~~
/home/simple/code/saas/ae-arbiter/src/grpcxx/predictor.grpcxx.pb.h:17:43: note: suggested alternatives:
In file included from /home/simple/code/saas/ae-arbiter/src/grpcxx/predictor.pb.h:33,
                 from /home/simple/code/saas/ae-arbiter/src/grpcxx/predictor.grpcxx.pb.h:11,
                 from /home/simple/code/saas/ae-arbiter/executables/yakoo-server.cc:7:
/opt/online/include/google/protobuf/empty.pb.h:60:23: note:   ‘google::protobuf::Empty’
   60 | class PROTOBUF_EXPORT Empty final :
      |                       ^~~~~
In file included from /opt/online/include/folly/futures/Promise.h:23,
                 from /opt/online/include/folly/futures/Future.h:39,
                 from /home/simple/code/saas/ae-arbiter/executables/yakoo-server.cc:6:
/opt/online/include/folly/futures/detail/Core.h:52:3: note:   ‘folly::futures::detail::State::Empty’
   52 |   Empty = 1 << 6,
      |   ^~~~~
In file included from /home/simple/code/saas/ae-arbiter/executables/yakoo-server.cc:7:
/home/simple/code/saas/ae-arbiter/src/grpcxx/predictor.grpcxx.pb.h:17:65: error: template argument 2 is invalid
   17 | using rpcversion = grpcxx::rpc<"version", Empty, VersionResponse>;
      |                                                                 ^
/home/simple/code/saas/ae-arbiter/src/grpcxx/predictor.grpcxx.pb.h:17:28: error: ‘<expression error>’ in namespace ‘grpcxx’ does not name a type
   17 | using rpcversion = grpcxx::rpc<"version", Empty, VersionResponse>;
      |                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/simple/code/saas/ae-arbiter/src/grpcxx/predictor.grpcxx.pb.h:19:69: error: ‘rpcversion’ was not declared in this scope
   19 | using Service = grpcxx::service<"ifastverse.Predictor", rpcpredict, rpcversion>;
      |                                                                     ^~~~~~~~~~
/home/simple/code/saas/ae-arbiter/src/grpcxx/predictor.grpcxx.pb.h:19:79: error: template argument 3 is invalid
   19 | using Service = grpcxx::service<"ifastverse.Predictor", rpcpredict, rpcversion>;
      |                                                                               ^
/home/simple/code/saas/ae-arbiter/src/grpcxx/predictor.grpcxx.pb.h:19:25: error: ‘<expression error>’ in namespace ‘grpcxx’ does not name a type
   19 | using Service = grpcxx::service<"ifastverse.Predictor", rpcpredict, rpcversion>;
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/simple/code/saas/ae-arbiter/executables/yakoo-server.cc:10:1: error: expected unqualified-id before string constant

--- a/src/protoc-gen-grpcxx/grpcxx.cpp
+++ b/src/protoc-gen-grpcxx/grpcxx.cpp
@@ -50,11 +50,30 @@ bool Grpcxx::Generate(
 		for (int j = 0; j < service->method_count(); j++) {
 			auto method = service->method(j);
 
+			// Map proto types to c++ types
+			// e.g. `google.protobuf.Empty` -> `google::protobuf::Empty`
+			auto mapper = [&file](const google::protobuf::Descriptor *t) -> std::string {
+				if (t->full_name().starts_with(file->package())) {
+					return t->name();
+				}
+
+				std::string result;
+				std::size_t start = 0;
+				for (auto end = t->full_name().find('.'); end != std::string::npos;
+					 end      = t->full_name().find('.', start)) {
+					result += t->full_name().substr(start, end - start) + "::";
+					start   = end + 1;
+				}
+				result += t->full_name().substr(start);
+
+				return result;
+			};
+
 			output += fmt::format(
 				"using rpc{0} = grpcxx::rpc<\"{0}\", {1}, {2}>;\n\n",
 				method->name(),
-				method->input_type()->name(),
-				method->output_type()->name());
+				mapper(method->input_type()),
+				mapper(method->output_type()));
 
 			tmp += fmt::format(", rpc{}", method->name());
 		}

--- a/src/protoc-gen-grpcxx/grpcxx.cpp
+++ b/src/protoc-gen-grpcxx/grpcxx.cpp
@@ -53,18 +53,19 @@ bool Grpcxx::Generate(
 			// Map proto types to c++ types
 			// e.g. `google.protobuf.Empty` -> `google::protobuf::Empty`
 			auto mapper = [&file](const google::protobuf::Descriptor *t) -> std::string {
-				if (t->full_name().starts_with(file->package())) {
-					return t->name();
+				auto name = t->full_name();
+				if (name.starts_with(file->package())) {
+					name = name.substr(file->package().size() + 1);
 				}
 
 				std::string result;
 				std::size_t start = 0;
-				for (auto end = t->full_name().find('.'); end != std::string::npos;
-					 end      = t->full_name().find('.', start)) {
-					result += t->full_name().substr(start, end - start) + "::";
+				for (auto end = name.find('.'); end != std::string::npos;
+					 end      = name.find('.', start)) {
+					result += name.substr(start, end - start) + "::";
 					start   = end + 1;
 				}
-				result += t->full_name().substr(start);
+				result += name.substr(start);
 
 				return result;
 			};

